tmk/cpp/bin/tmk-query-with-faiss.cpp

```diff
@@ -161,23 +164,36 @@ int main(int argc, char** argv) {
   size_t num_database_vectors = haystackMetadataToFeatures.size();
 
   // Make the index object and train it
-  faiss::IndexFlatL2 coarse_quantizer(vector_dimension);
+  faiss::IndexFlatIP coarse_quantizer(vector_dimension);
 
   // A reasonable number of centroids to index num_database_vectors vectors
   int num_centroids = int(4 * sqrt(num_database_vectors));
   if (num_centroids > num_database_vectors) {
     num_centroids = num_database_vectors;
   }
-  printf("VECTOR_DIMENSION     %d\n", (int)vector_dimension);
-  printf("NUM_DATABASE_VECTORS %d\n", (int)num_database_vectors);
-  printf("NUM_CENTROIDS        %d\n", (int)num_centroids);
+  if (verbose) {
+    printf("VECTOR_DIMENSION     %d\n", (int)vector_dimension);
+    printf("NUM_DATABASE_VECTORS %d\n", (int)num_database_vectors);
+    printf("NUM_CENTROIDS        %d\n", (int)num_centroids);
+  }
 
   // The coarse quantizer should not be deallocated before the index.
   // 4 = number of bytes per code (vector_dimension must be a multiple of this)
   // 8 = number of bits per sub-code (almost always 8)
   faiss::IndexIVFPQ faiss_index(&coarse_quantizer, vector_dimension, num_centroids, 4, 8);
   faiss_index.verbose = verbose;
 
+  // check for min data size for optimal FAISS performance
+  int min_points = faiss_index.cp.min_points_per_centroid * num_centroids;
+  if (num_database_vectors < min_points) {
+    if (verbose) {
+      printf("MINIMUM POINTS NOT MET FOR FAISS: %d FOUND vs %d EXPECTED\n", (int)num_database_vectors, min_points);
+    }
+    if (!force) {
+      return tmkQuery(argc, argv);
```

Comment:
(blocking): This seems spooky to me - tmkQuery might have a different set of arugments and parsing function, and the error messages seem like they'll get confusing, since they each have their own usage() function.

Additionally, this file is called "query with faiss" and so I'd expect to it query with faiss or just error out, without fallback behavior.